awx_collection/plugins/modules/tower_inventory.py

```diff
@@ -106,31 +106,37 @@ def main():
     kind = module.params.get('kind')
     host_filter = module.params.get('host_filter')
 
-    json_output = {'inventory': name, 'state': state}
-
-    tower_auth = tower_auth_config(module)
-    with settings.runtime_values(**tower_auth):
-        tower_check_mode(module)
-        inventory = tower_cli.get_resource('inventory')
-
-        try:
-            org_res = tower_cli.get_resource('organization')
-            org = org_res.get(name=organization)
-
-            if state == 'present':
-                result = inventory.modify(name=name, organization=org['id'], variables=variables,
-                                          description=description, kind=kind, host_filter=host_filter,
-                                          create_on_missing=True)
-                json_output['id'] = result['id']
-            elif state == 'absent':
-                result = inventory.delete(name=name, organization=org['id'])
-        except (exc.NotFound) as excinfo:
-            module.fail_json(msg='Failed to update inventory, organization not found: {0}'.format(excinfo), changed=False)
-        except (exc.ConnectionError, exc.BadRequest, exc.AuthError) as excinfo:
-            module.fail_json(msg='Failed to update inventory: {0}'.format(excinfo), changed=False)
-
-    json_output['changed'] = result['changed']
-    module.exit_json(**json_output)
+    # Attempt to look up the related items the user specified (these will fail the module if not found)
+    org_id = module.resolve_name_to_id('organizations', organization)
+
+    # Attempt to look up inventory based on the provided name and org ID
+    inventory = module.get_one('inventories', **{
+        'data': {
+            'name': name,
+            'organization': org_id
+        }
+    })
+
+    # Create data to sent to create and update
+    inventory_fields = {
+        'name': name,
+        'description': description,
+        'organization': org_id,
+        'variables': variables,
+        'kind': kind,
+        'host_filter': host_filter,
+    }
+
+    if state == 'absent':
+        # If the state was absent we can let the module delete it if needed, the module will handle exiting from this
+        module.delete_if_needed(inventory)
+    elif state == 'present':
+        # We need to perform a check to make sure you are not trying to convert a regular inventory into a smart one.
+        if inventory and inventory['kind'] == '' and inventory_fields['kind'] == 'smart':
+            module.fail_json(msg='You cannot turn a regular inventory into a "smart" inventory.')
```

Comment:
Won't the API itself give a 400 in this event? If it does that, and if the 400 error is handled correctly, then I don't think we should need this.

that case can be unit tested. As long as it gives a sensible error message that says mostly the same thing, we shouldn't need this conditional.

Reply:
The API returns a `405` without a helpful error message:

![Screen Shot 2020-02-10 at 9 32 16 AM](https://user-images.githubusercontent.com/28930622/74158617-63f94d80-4be8-11ea-87fb-a6f9a35ac64a.png)

I filed a bug for this here: https://github.com/ansible/awx/issues/5744